File: backend/internal/domain/repository/repository.go
package repository

import (
	"time"
)

// Repository represents a Git repository configuration
// Self-contained design: repository stores all necessary info, no git_provider_id dependency
type Repository struct {
	ID             int64 `gorm:"primaryKey" json:"id"`
	OrganizationID int64 `gorm:"not null;index" json:"organization_id"`

	// Provider info (self-contained, no foreign key to git_providers)
	ProviderType    string `gorm:"size:50;not null" json:"provider_type"`      // github, gitlab, gitee, generic
	ProviderBaseURL string `gorm:"size:255;not null" json:"provider_base_url"` // https://github.com, https://gitlab.company.com
	CloneURL        string `gorm:"size:500" json:"clone_url"`                  // Full clone URL

	ExternalID    string  `gorm:"size:255;not null" json:"external_id"` // Provider's project ID
	Name          string  `gorm:"size:255;not null" json:"name"`
	FullPath      string  `gorm:"size:500;not null" json:"full_path"`
	DefaultBranch string  `gorm:"size:100;default:'main'" json:"default_branch"`
	TicketPrefix  *string `gorm:"size:10" json:"ticket_prefix,omitempty"` // Ticket prefix like 'AM'

	// Visibility: "organization" (all members can see), "private" (only importer can see)
	Visibility       string `gorm:"size:20;not null;default:'organization'" json:"visibility"`
	ImportedByUserID *int64 `gorm:"index" json:"imported_by_user_id,omitempty"` // User who imported this repo

	IsActive bool `gorm:"not null;default:true" json:"is_active"`

	CreatedAt time.Time  `gorm:"not null;default:now()" json:"created_at"`
	UpdatedAt time.Time  `gorm:"not null;default:now()" json:"updated_at"`
	DeletedAt *time.Time `gorm:"index" json:"deleted_at,omitempty"` // Soft delete support
}

func (Repository) TableName() string {
	return "repositories"
}

// Branch represents a Git branch
type Branch struct {
	Name      string `json:"name"`
	IsDefault bool   `json:"is_default"`
	Commit    string `json:"commit,omitempty"`
}

// WebhookConfig represents webhook configuration for a repository
type WebhookConfig struct {
	ID        string   `json:"id"`
	URL       string   `json:"url"`
	Events    []string `json:"events"`
	IsActive  bool     `json:"is_active"`
	CreatedAt string   `json:"created_at,omitempty"`
}

File: web/src/lib/api/repository.ts
import { request } from "./base";

// Repository types (self-contained, no git_provider_id)
export interface RepositoryData {
  id: number;
  organization_id: number;
  provider_type: string; // github, gitlab, gitee, generic
  provider_base_url: string; // https://github.com
  clone_url: string;
  external_id: string;
  name: string;
  full_path: string;
  default_branch: string;
  ticket_prefix?: string;
  visibility: string; // "organization" or "private"
  imported_by_user_id?: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreateRepositoryRequest {
  provider_type: string;
  provider_base_url: string;
  clone_url?: string;
  external_id: string;
  name: string;
  full_path: string;
  default_branch?: string;
  ticket_prefix?: string;
  visibility?: string;
}

export interface UpdateRepositoryRequest {
  name?: string;
  default_branch?: string;
  ticket_prefix?: string;
  is_active?: boolean;
}

// Repository API
export const repositoryApi = {
  list: () => {
    return request<{ repositories: RepositoryData[] }>(`/api/v1/org/repositories`);
  },

  get: (id: number) =>
    request<{ repository: RepositoryData }>(`/api/v1/org/repositories/${id}`),

  create: (data: CreateRepositoryRequest) =>
    request<{ repository: RepositoryData }>("/api/v1/org/repositories", {
      method: "POST",
      body: data,
    }),

  update: (id: number, data: UpdateRepositoryRequest) =>
    request<{ repository: RepositoryData }>(`/api/v1/org/repositories/${id}`, {
      method: "PUT",
      body: data,
    }),

  delete: (id: number) =>
    request<{ message: string }>(`/api/v1/org/repositories/${id}`, {
      method: "DELETE",
    }),

  listBranches: (id: number, accessToken: string) =>
    request<{ branches: string[] }>(`/api/v1/org/repositories/${id}/branches`, {
      headers: { "X-Git-Access-Token": accessToken },
    }),

  syncBranches: (id: number, accessToken: string) =>
    request<{ branches: string[] }>(`/api/v1/org/repositories/${id}/sync-branches`, {
      method: "POST",
      body: { access_token: accessToken },
    }),

  setupWebhook: (id: number) =>
    request<{ message: string; webhook_url?: string }>(`/api/v1/org/repositories/${id}/webhook`, {
      method: "POST",
    }),
};
